File: src/engines/observation.py
"""Private player knowledge and immutable input shared by engine policies."""

from collections.abc import Callable
from dataclasses import dataclass
from random import Random

from bluff_mau_mau import Card, Challenge, GameState, Move, PlayCard


PileKnowledge = tuple[frozenset[Card], frozenset[Card]]
EMPTY_KNOWLEDGE: PileKnowledge = (frozenset(), frozenset())


@dataclass(frozen=True, slots=True)
class Observation:
    player: int
    hand: tuple[Card, ...]
    opponent_count: int
    top: Card
    chosen_suit: str | None
    phase: str
    draw_penalty: int
    skip_pending: bool
    provisional_winner: int | None
    deck_count: int
    pile_count: int
    known_pile_cards: frozenset[Card] = frozenset()

    def __post_init__(self):
        object.__setattr__(self, "hand", tuple(self.hand))
        object.__setattr__(self, "known_pile_cards", frozenset(self.known_pile_cards))


Bot = Callable[[Observation, tuple[Move, ...], Random], Move]


def _knowledge_pair(knowledge) -> PileKnowledge:
    try:
        first, second = knowledge
        pair = (frozenset(first), frozenset(second))
    except (TypeError, ValueError) as error:
        raise ValueError("Knowledge needs two sets of actual cards") from error
    if any(type(card) is not Card for known in pair for card in known):
        raise ValueError("Knowledge needs concrete Card identities")
    return pair


def new_knowledge(starting_card: Card) -> PileKnowledge:
    """Both players see the starting discard when a new game is dealt."""
    if type(starting_card) is not Card:
        raise ValueError("The starting identity must be a Card")
    known = frozenset((starting_card,))
    return known, known


def advance_knowledge(before: GameState, move: Move, after: GameState,
                      knowledge: PileKnowledge) -> PileKnowledge:
    """Track one genuine Play transition without revealing accepted hidden cards.

    The trusted runner has already validated/applied the move. Intersecting only
    previously known or newly revealed identities with the remaining pile removes
    recycled cards; it never adds knowledge from the hidden pile itself.
    """
    known = list(_knowledge_pair(knowledge))
    if isinstance(move, PlayCard):
        known[before.turn] = known[before.turn] | {move.actual_card}
    elif isinstance(move, Challenge):
        revealed = {before.pile[-1]}
        known = [cards | revealed for cards in known]
    remaining = frozenset(after.pile)
    return known[0] & remaining, known[1] & remaining


def observe(state: GameState, knowledge: PileKnowledge = EMPTY_KNOWLEDGE) -> Observation:
    """Project a valid game state and coherent history for the next actor only."""
    known = _knowledge_pair(knowledge)
    return Observation(
        player=state.turn, hand=state.hands[state.turn],
        opponent_count=len(state.hands[1 - state.turn]), top=state.top,
        chosen_suit=state.chosen_suit, phase=state.phase,
        draw_penalty=state.draw_penalty, skip_pending=state.skip_pending,
        provisional_winner=state.provisional_winner,
        deck_count=len(state.deck), pile_count=len(state.pile),
        known_pile_cards=known[state.turn],
    )

File: src-rust/src/engines/observation.rs
//! Only facts the acting player can know cross the policy boundary.
use crate::game::{CARDS, Card, GameState, Move, Phase, Suit};
use crate::rng::PythonRandom;

pub type PileKnowledge = [u32; 2];
pub const EMPTY_KNOWLEDGE: PileKnowledge = [0, 0];

#[derive(Clone, Debug, PartialEq)]
pub struct Observation {
    pub player: usize,
    pub hand: Vec<Card>,
    pub opponent_count: usize,
    pub top: Card,
    pub chosen_suit: Option<Suit>,
    pub phase: Phase,
    pub draw_penalty: u32,
    pub skip_pending: bool,
    pub provisional_winner: Option<usize>,
    pub deck_count: usize,
    pub pile_count: usize,
    /// A 32-bit set of known actual card identities, never their shuffled order.
    pub known_pile_cards: u32,
    pub opening_card: bool,
}

pub trait Bot: Send + Sync {
    fn name(&self) -> String;
    fn choose(
        &self,
        observation: &Observation,
        legal_moves: &[Move],
        rng: &mut PythonRandom,
    ) -> Result<Move, String>;
}

pub fn card_mask(card: Card) -> u32 {
    1u32 << card.0
}
pub fn known_cards(mask: u32) -> Vec<Card> {
    CARDS
        .into_iter()
        .filter(|card| mask & card_mask(*card) != 0)
        .collect()
}
pub fn new_knowledge(starting_card: Card) -> PileKnowledge {
    [card_mask(starting_card); 2]
}

pub fn advance_knowledge(
    before: &GameState,
    action: &Move,
    after: &GameState,
    knowledge: PileKnowledge,
) -> PileKnowledge {
    advance_known(
        before.turn,
        *before.pile.last().unwrap(),
        action,
        &after.pile,
        knowledge,
    )
}

/// The trusted match loop needs only the old actor and old top, not a full state clone.
pub(crate) fn advance_known(
    actor: usize,
    old_actual: Card,
    action: &Move,
    remaining_pile: &[Card],
    mut knowledge: PileKnowledge,
) -> PileKnowledge {
    match action {
        Move::Play { actual_card, .. } => knowledge[actor] |= card_mask(*actual_card),
        Move::Challenge => {
            knowledge[0] |= card_mask(old_actual);
            knowledge[1] |= card_mask(old_actual);
        }
        _ => {}
    }
    let remaining = remaining_pile
        .iter()
        .fold(0, |mask, card| mask | card_mask(*card));
    [knowledge[0] & remaining, knowledge[1] & remaining]
}

pub fn observe(state: &GameState, knowledge: PileKnowledge) -> Observation {
    Observation {
        player: state.turn,
        hand: state.hands[state.turn].clone(),
        opponent_count: state.hands[1 - state.turn].len(),
        top: state.top,
        chosen_suit: state.chosen_suit,
        phase: state.phase,
        draw_penalty: state.draw_penalty,
        skip_pending: state.skip_pending,
        provisional_winner: state.provisional_winner,
        deck_count: state.deck.len(),
        pile_count: state.pile.len(),
        known_pile_cards: knowledge[state.turn],
        opening_card: state.opening_card,
    }
}
